Managed Upgrade Operator

hawkowl opened the pull request:
### Which issue this PR addresses:

https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/10430647

### What this PR does / why we need it:

Implements the Managed Upgrade Operator deployment in the ARO Operator.

### Test plan for issue:

Unit tests exist.

To verify, build the operator and push it to a cluster, and then perform an AdminUpdate with this:

```
{
  "properties": {
    "maintenanceTask": "OperatorUpdate",
    "operatorFlags": {
      "rh.srep.muo.enabled": "true",
      "rh.srep.muo.managed": "true",
      "rh.srep.muo.deploy.pullspec": "quay.io/app-sre/managed-upgrade-operator@sha256:054895ba677031579ed346ad9df7ae7f9c6b5ed9ca106f34fb028bbc2aee4ff4"
    }
  }
}
```

Then apply this config manifest (will be included in a separate PR with configuration options):

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: managed-upgrade-operator-config
  namespace:  openshift-managed-upgrade-operator
data:
  config.yaml: |
    configManager:
      source: LOCAL
      localConfigName: managed-upgrade-config
      watchInterval: 1
    maintenance:
      controlPlaneTime: 90
      ignoredAlerts:
        controlPlaneCriticals:
        - ClusterOperatorDown
        - ClusterOperatorDegraded
    upgradeWindow:
      delayTrigger: 30
      timeOut: 120
    nodeDrain:
      timeOut: 45
      expectedNodeDrainTime: 8
    scale:
      timeOut: 30
    healthCheck:
      ignoredCriticals:
      - PrometheusRuleFailures
      - CannotRetrieveUpdates
      - FluentdNodeDown
      ignoredNamespaces:
      - openshift-logging
      - openshift-redhat-marketplace
      - openshift-operators
      - openshift-user-workload-monitoring
      - openshift-pipelines

```

You may need to kill the MUO pod for it to pick up the config.

You can then schedule an upgrade:

```
apiVersion: upgrade.managed.openshift.io/v1alpha1
kind: UpgradeConfig
metadata:
  name: managed-upgrade-config
  namespace: openshift-managed-upgrade-operator
spec:
  type: "ARO"
  upgradeAt: "2022-02-08T03:20:00Z"
  PDBForceDrainTimeout: 60
  capacityReservation: false
  desired:
    image: "<IMAGEHASH HERE>"

```

### Is there any documentation that needs to be updated for this PR?

Customer-facing and SRE-facing documentation TBA.

m1kola commented on pkg/operator/controllers/muo/muo_controller.go:
```diff
@@ -0,0 +1,97 @@
+)
+
+const (
+	CONFIG_NAMESPACE string = "rh.srep.muo"
+	ENABLED          string = CONFIG_NAMESPACE + ".enabled"
+	MANAGED          string = CONFIG_NAMESPACE + ".managed"
+	PULLSPEC         string = CONFIG_NAMESPACE + ".pullspec"
```
* First upper case char makes this const exported which seems unnecessary as it is not being used outside of the package
* We use camel case

We need to address this across all controllers (not in this PR, but in general).

hawkowl replied:
I ideally wanted to be able to use them in things like the E2E tests.

Also, are global constants not capital snake case, or is that just a C thing?

m1kola commented on pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:
```diff
@@ -73,6 +73,11 @@ func (f OperatorFlags) GetSimpleBoolean(key string) bool {
 	return false
 }
 
+func (f OperatorFlags) Get(key string) (val string, exists bool) {
+	v, ext := f[key]
+	return v, ext
+}
```
So it is not just flags anymore... We use it as a key/value now?

hawkowl replied:
This is used for two purposes:

1. Where I want to check for a null value (because explicit true adds, explicit false removes, incorrect or undefined does nothing as per the feature flags doc)
2. Defining the pull spec for the container, which we want to avoid bundling into the RP (e.g. to avoid situations such as the fluentbit version situation).

(In general: The lack of schema is to avoid backwards/forwards compatibility issues between the RP, Operator, etc, so unfortunately encoding something that limits the uses so we don't misuse them more difficult.)

hawkowl commented:
I re-verified the instructions today and it works :)

m1kola commented on pkg/operator/controllers/muo/deploy.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+func (o *muoDeployer) Remove(ctx context.Context) error {
+	for _, assetName := range AssetNames() {
```
Do we care about the order? `AssetNames()` returns a slice which is converted from a `map`. Iterating over a map in go is randomised.

m1kola replied:
I think, it is not a problem in `CreateOrUpdate` becasue `dynamichelper.Prepare(resources)` deals with the order as far as I can see, but can it cause issues with deletion?

hawkowl replied:
If we want to delete the whole thing, yes, but right now we are only deleting a single deployment, so I didn't bother.

m1kola commented on pkg/operator/controllers/muo/muo_controller_test.go:
```diff
@@ -0,0 +1,180 @@
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```
This is really hard to follow :( I think this is partially because we combine tests of (at least) two units here:

* We test `Reconcile` and it's logic related to flags
* And we test deployer

IMO in `muo_controller_test.go` we should have a simple(ish) test (probably just one for `Reconcile`) where we have a mock of a deployer and test things like:

* What if controller is disabled
* What if `managed` is `"true"`
  * What if `deployer.CreateOrUpdate` returns `err != nil`?
  * What if `deployer.CreateOrUpdate` returns `err == nil`?
* What if `managed` is `"false"`
  * What if `deployer.Remove` returns `err != nil`?
  * What if `deployer.Remove` returns `err == nil`?
* What if `managed` is `""`
* Something else?

And we should also have `deploy_test.go`. In this one we should mock dynamic helper (maybe something else?).

I think it will be enough to check that `CreateOrUpdate` (one test):
* Creates right resource types (so we don't duplicate yamls in tests)
* Check that we replace image with the right pull spec

And check that `Remove` (a separate test, I think):

* Removes the right resource types

Everything else should be tested inside the dynamic helper and is out of scope for this.

---

Sorry for the brain dump. Does any of this make sense?

hawkowl replied:
I think I've implemented most of this, now.

m1kola replied:
It is now a bit better in terms of readability: we no longer test deployer via this test. But it is still very hard to follow. I spent more than an hour trying to understand this test and what cases it covers.

What makes it hard in my opition:

1. A lot of flags/field in each test case with a lot of permutations. We are trying to reduce code duplication, but it hurts readability.
2. Mocking dynamic helper instead of mocking deployer

The above means that it is very time consuming to add a new test case because you need to find a right combination of test case flags. And it is hard to follow which calls are expected in which test case. 

And we need to jump between too many things to understand the test: test case struct, test body, between `Reconciler` implementation (subject of the test) and between the deployer (which should not be our concern here at all as it is tested separately) as well as `ready.CheckDeploymentIsReady` (also not our concern - it is like 2 steps down from what we are actually testing).

IMO, we should introduce an interface for deployer and mock it. I belive we will have to make it exportable to be able to generate a mock, but it is a small price to pay for readability.

I think we can reduce the test case to the following fields:

* Test case name
* A function which sets up mocks for each test case
* Probably, a function that modifies `arov1alpha1.Cluster` in a way we want before the test
* Wanted err message `string`

And the test body will be things common to al ltest cases

m1kola reviewed (changes requested):
Thanks Amber. I think this is on the right track.

I left quite a lof of comments and here I'll try to give a summary of things that bother me the most (more details are below):

* Controller-runtime catch: `RequeueAfter` and `err` logic is not what you intended, I think.
* We only seem to reconcile resources in response to ` arov1alpha1.Cluster` events. I think we need to own MUO objects and reconcile them if they change.
* Complexity and the scope of the tests
* Potential issues in `muoDeployer.Remove` with deletion order.

Rest of the comments are minor things.

s-amann commented on pkg/operator/controllers/muo/deploy.go:
```diff
@@ -0,0 +1,115 @@
+type Interface interface {
+	CreateOrUpdate(context.Context) error
+	Remove(context.Context) error
+	//IsReady(context.Context) (bool, error)
```
nit: can we delete the commented code if the above is true?

hawkowl replied:
I decided to implement the check that gets exposed in statuses, since we don't otherwise have a nice way of getting MUO's state right now.

hawkowl commented on pkg/operator/controllers/muo/muo_controller.go:
```diff
@@ -0,0 +1,97 @@
+		return o.GetName() == arov1alpha1.SingletonClusterName
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&arov1alpha1.Cluster{}, builder.WithPredicates(aroClusterPredicate)).
+		Named(controllers.ManagedUpgradeOperatorControllerName).
+		Complete(r)
+}
```
Perhaps? I honestly don't have my entire head around how this section works :/

m1kola replied:
Controller runtime has a bit [in docs](https://pkg.go.dev/sigs.k8s.io/controller-runtime#section-readme) on this (grep for "owner" and "owns").

And [operator SDK docs](https://sdk.operatorframework.io/docs/) have some useful info about this stuff too. We do not use operator SDK anymore, but it is based on controller runtime and docs a little bit better.

hawkowl replied:
I tried to make this a bit better, although I have no idea if the way I did it is awful or not (I think it is :) )

hawkowl commented on pkg/operator/controllers/muo/muo_controller_test.go:
```diff
@@ -0,0 +1,180 @@
+		wantErr         bool
+	}{
+		{
+			name:    "disabled",
```
this should be covered now

hawkowl commented on pkg/operator/controllers/muo/muo_controller.go:
```diff
@@ -0,0 +1,97 @@
+)
+
+type Reconciler struct {
+	log *logrus.Entry
```
I think that if the DynamicHelper fails, it logs... that was what was happening in dev, at least?

m1kola commented on pkg/operator/controllers/muo/deploy_test.go:
```diff
@@ -0,0 +1,126 @@
+}
+
+func TestDeployDelete(t *testing.T) {
+	controller := gomock.NewController(t)
+	defer controller.Finish()
+
+	cluster := &arov1alpha1.Cluster{}
+	k8scli := fake.NewSimpleClientset()
+
+	dh := mock_dynamichelper.NewMockInterface(controller)
+
+	// keep track of deleted objects
+	deleted := []string{}
+	checkDelete := func(ctx context.Context, kind string, namespace string, name string) error {
+		deleted = append(deleted, fmt.Sprintf("%s/%s/%s", kind, namespace, name))
+		return nil
+	}
+	dh.EXPECT().EnsureDeleted(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).AnyTimes().Do(checkDelete).Return(nil)
+
+	deployer := NewMUODeployer(cluster, k8scli, dh)
+	errs := deployer.Remove(context.Background())
+	if len(errs) != 0 {
+		t.Error(errs)
+	}
+
+	expectedDeleted := []string{
+		"Deployment/openshift-managed-upgrade-operator/managed-upgrade-operator",
+	}
+	sort.Strings(deleted)
+	sort.Strings(expectedDeleted)
+	mismatches := deep.Equal(deleted, expectedDeleted)
+	for _, e := range mismatches {
+		t.Error(e)
+	}
```
Go mock is great in simple cases like this ;)

```suggestion
	controller := gomock.NewController(t)
	defer controller.Finish()

	cluster := &arov1alpha1.Cluster{}
	k8scli := fake.NewSimpleClientset()

	dh := mock_dynamichelper.NewMockInterface(controller)
	dh.EXPECT().EnsureDeleted(gomock.Any(), "Deployment", "openshift-managed-upgrade-operator", "managed-upgrade-operator").Return(nil)

	deployer := NewMUODeployer(cluster, k8scli, dh)
	errs := deployer.Remove(context.Background())
	if len(errs) != 0 {
		t.Error(errs)
	}
```

m1kola commented on pkg/operator/controllers/muo/deploy_test.go:
```diff
@@ -0,0 +1,126 @@
+			if d, ok := i.(*appsv1.Deployment); ok {
+				deployments = append(deployments, d)
+			}
+			deployedObjects[kind] = true
```
We deploy multiple roles. We might accidentaly deelete one of the role yamls and not catch it.

Maybe let's use a slice of kinds? This way we make sure that we deploy the right number of objects of right kinds, but don't go in to too much details. Trying to find a balance here. What do you think?

hawkowl replied:
I went with a map of kind: count, which seems simple and clear. Considering making sure we put things in the right namespaces, but I think that's maybe out of scope of this (since we may well need to put things in various namespaces if MUO needs it)

m1kola commented on pkg/operator/controllers/muo/muo_controller.go:
```diff
@@ -0,0 +1,161 @@
+		timeoutCtx, cancel := context.WithTimeout(ctx, r.readinessTimeout)
+		defer cancel()
+
+		err := wait.PollImmediateUntil(r.readinessPollTime, func() (bool, error) {
+			return deployer.IsReady(ctx)
+		}, timeoutCtx.Done())
+
+		if err != nil {
+			condition = &operatorv1.OperatorCondition{
+				Type:    arov1alpha1.ManagedUpgradeOperatorStatus,
+				Status:  operatorv1.ConditionFalse,
+				Message: "Deployment did not become ready",
+				Reason:  "DeploymentFailed",
+			}
+		} else {
+			condition = &operatorv1.OperatorCondition{
+				Type:    arov1alpha1.ManagedUpgradeOperatorStatus,
+				Status:  operatorv1.ConditionTrue,
+				Message: "Deployment is ready",
+				Reason:  "DeploymentSucceeded",
+			}
+		}
```
Do we really care about exposing these conditions?

---

What do you think if we do the following?

In the controller
* Wait for the deployment to become ready after `deployer.CreateOrUpdate`.
  * If we error waiting (currently line 98) for deployment - return this error to requeue.
  * If ready - all good
 * Do not wait after `deployer.Remove`

This way we make sure that we do not try to create or update multiple times while containers are still creating.

In the monitor:
* Implement something like [ARO operator heartbeat](https://github.com/Azure/ARO-RP/blob/efe990b0a39f7f4889d452da99f38488e4c6c1e1/pkg/monitor/cluster/arooperatorheartbeat.go), but for MUO.
* Potentially create an alert on it

This way we monitor clusters with enabled and managed MUO operator and do not monitor everything else... And we don't have to deal with a mess of conditions.

If after `deployer.CreateOrUpdate` dpeloyment doesn't become ready - we should pick this up with the monitor. Should we care about any other cases?

hawkowl replied:
Fair -- I am not super sure how we want to monitor all of this yet, since there's PrometheusRule stuff from OSD that we can't use that triggers Prometheus alerts. I've removed it for now, we can implement a better version before full GA.

m1kola commented on pkg/operator/controllers/muo/muo_controller.go:
```diff
@@ -0,0 +1,161 @@
+		}
+	} else if strings.EqualFold(managed, "false") {
+		errs := deployer.Remove(ctx)
+		if len(errs) != 0 {
```
See the comment above about getting rid of the conditions.

But if we decide that we actually want conditions - we should still requeue in case of errors on `deployer.Remove` to do a proper cleanup.

hawkowl commented on pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:
```diff
@@ -73,6 +73,11 @@ func (f OperatorFlags) GetSimpleBoolean(key string) bool {
 }
 
+func (f OperatorFlags) Get(key string) (val string, exists bool) {
+	v, ext := f[key]
```
removed this :)

s-amann commented on pkg/env/dev.go:
```diff
@@ -74,7 +74,12 @@ func (d *dev) AROOperatorImage() string {
 	}
 
-	return fmt.Sprintf("%s/aro:%s", d.ACRDomain(), version.GitCommit)
+	operatorAcr := os.Getenv("ARO_OPERATOR_ACR")
+	if operatorAcr == "" {
+		operatorAcr = d.ACRDomain()
+	}
+
+	return fmt.Sprintf("%s/aro:%s", operatorAcr, version.GitCommit)
```
I don't see `ARO_OPERATOR_ACR` mentioned elsewhere in the code base. Is this additional env var to help format the image name?

Why wouldn't we put this format into the value of `ARO_IMAGE`?

hawkowl replied:
Added a comment in the docs. I thought the option was a good idea, considering the iterative changes I was making on the Operator.

hawkowl replied:
ARO_IMAGE is a static image -- e.g. if you want to use the prebuilt one. ARO_OPERATOR_ACR sets a custom Docker repo as the base, that then has the hash put on it.

We don't want to set the global ACR_RESOURCE_ID to it because then things like fluentbit, MDSD, etc, will look at that. We want to use the int ACR for all of those resources, but our custom ACR specifically for the Operator.

rogbas commented on pkg/operator/controllers/muo/deploy.go:
```diff
@@ -0,0 +1,96 @@
+
+	for _, obj := range resources {
+		// delete any deployments we have
+		if d, ok := obj.(*appsv1.Deployment); ok {
```
Why are we caring only about `deployment` cleanup in the Remove step? Shouldn't we also remove Roles/Bindings/Service Accounts created with the MUO manifest?

hawkowl replied:
Right now the priority is "have MUO not run", full cleanup and ordering is something complicated enough that I elided it for this RP release.

rogbas replied:
yes, IMO It sounds like a good story for a follow-up epic.

m1kola commented on pkg/operator/controllers/muo/muo_controller.go:
```diff
@@ -0,0 +1,135 @@
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&arov1alpha1.Cluster{}, builder.WithPredicates(aroClusterPredicate))
+
+	resources, err := r.deployer.resources("")
+	if err != nil {
+		return err
+	}
+
+	for _, i := range resources {
+		o, ok := i.(client.Object)
+		if ok {
+			builder.Owns(o)
+		}
+	}
```
1. I'm not sure that this will work. `deployer.resources` returns a slice of `kruntime.Object`, not a specific type. Will controller-runtime be able to figure out a specific type/kind from it?
2. Where do we set ownership for these objects we deploy?

---

We need careful testing here:

1. We need to make sure that objects we deploy have the ownership set properly.
2. We need to make sure that making a change on objects we deploy - will trigger the reconciler.

hawkowl replied:
In the logs when I tested it, I got this:

```
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: /, Kind="
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=RoleBinding"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=Role"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=ClusterRoleBinding"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=Role"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: /v1, Kind=Namespace"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=RoleBinding"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=ClusterRole"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=RoleBinding"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=Role"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: /v1, Kind=ServiceAccount"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=RoleBinding"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: apps/v1, Kind=Deployment"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: rbac.authorization.k8s.io/v1, Kind=Role"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: /v1, Kind=ConfigMap"
time="2022-02-17T06:54:44Z" level=info msg="Starting EventSource" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:165" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster source="kind source: apiextensions.k8s.io/v1, Kind=CustomResourceDefinition"
time="2022-02-17T06:54:44Z" level=info msg="Starting Controller" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:173" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster
time="2022-02-17T06:54:44Z" level=info msg="Starting workers" func="controller.(*Controller).Start.func2()" file="vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:207" ManagedUpgradeOperator= controller= controller-runtime= manager= reconciler group=aro.openshift.io reconciler kind=Cluster worker count=1
```

I think that means it's working?

hawkowl replied:
Okay, I've now added setting ownership of objects we control, and a test for it.

m1kola commented on pkg/operator/controllers/muo/muo_controller.go:
```diff
@@ -0,0 +1,135 @@
+	} else if strings.EqualFold(managed, "false") {
+		errs := r.deployer.Remove(ctx)
+		if len(errs) != 0 {
+			r.log.Error("Error removing MUO:")
+			for _, e := range errs {
+				r.log.Error(e)
+			}
+			// just return one of the errors
+			return reconcile.Result{}, errs[0]
```
1. Multiple log calls will mean that it is likely that logs from this conroller will likely interleave with soemthing else. IMO, we need a single log call here for readability (if we want to log explicitly, see below).
2. Controller-runtime already does logging for us. So we will be logging all erorrs here and then controller runtime will log first error again.
3. Handling a slice of errors is tricky and not very useful in this case.

---

Taking above into account, I would like to suggest to change `Remove` signature to return a single error and inside `Remove`:

* Either concatenate all errors into a single error and return it.
* Or use error wrapping (https://pkg.go.dev/errors and https://go.dev/blog/go1.13-errors has a bit of info on wrapping)

And here we will jsut have our normal 

```go
	if err != nil {
		return reconcile.Result{}, err
	}
```

hawkowl replied:
Done

m1kola reviewed (changes requested):
I still have concerns about how we set ownership for MUO objects and reconcile them.

I haven't reviewed tests.

hawkowl commented on pkg/operator/controllers/muo/muo_controller.go:
```diff
@@ -0,0 +1,135 @@
+			return r.deployer.IsReady(ctx)
+		}, timeoutCtx.Done())
+		if err != nil {
+			r.log.Error("Managed Upgrade Operator deployment timed out on Ready")
+			return reconcile.Result{}, err
```
wrapped the err here

bryanro92 reviewed (commented):
Thanks for this PR. I've left a couple of comments.

hawkowl commented on pkg/operator/controllers/muo/deploy.go:
```diff
@@ -0,0 +1,119 @@
+
+		// set the image for the deployments
+		if d, ok := obj.(*appsv1.Deployment); ok {
+			if d.Labels == nil {
+				d.Labels = map[string]string{}
+			}
+
```
It is a copy paste :D 

I was originally setting the version here since it might have been useful, but ran into a length limitation (because the SHAs are full length coming from Quay, it seems).

hawkowl commented on pkg/operator/controllers/muo/muo_controller.go:
```diff
@@ -0,0 +1,128 @@
+		return reconcile.Result{}, nil
+	}
+
+	// the trusted-ca-bundle being injected shouldn't cause a reconcile
+	if request.Name == "trusted-ca-bundle" && request.Namespace == "openshift-managed-upgrade-operator" {
+		return reconcile.Result{}, nil
+	}
```
It's to prevent unneeded reconcilation.

Since OpenShift is responsible for maintaining this (by injecting the trust bundle), we don't want to respond to its reconciliation by OpenShift, I think.

hawkowl commented on pkg/operator/controllers/muo/muo_controller.go:
```diff
@@ -0,0 +1,128 @@
+			return r.deployer.IsReady(ctx)
+		}, timeoutCtx.Done())
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("Managed Upgrade Operator deployment timed out on Ready: %w", err)
```
does this rule also apply to names?

hawkowl commented on pkg/operator/controllers/muo/muo_controller.go:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	return builder.
+		WithEventFilter(predicate.Or(predicate.GenerationChangedPredicate{}, predicate.AnnotationChangedPredicate{})).
```
I know that OLM makes a label on the namespace, I feel like if we reconciled on that then they'd both fight forever :/

hawkowl commented on cmd/aro/operator.go:
```diff
@@ -191,6 +192,11 @@ func operator(ctx context.Context, log *logrus.Entry) error {
+		if err = (muo.NewReconciler(
+			log.WithField("controller", controllers.ManagedUpgradeOperatorControllerName),
+			arocli, kubernetescli, dh)).SetupWithManager(mgr); err != nil {
+			return fmt.Errorf("unable to create controller MachineSet: %v", err)
```
fixed

hawkowl commented on pkg/operator/controllers/muo/muo_controller_test.go:
```diff
@@ -0,0 +1,163 @@
+				return
+			} else if tt.wantErr && err.Error() != tt.errorString {
+				t.Errorf("Reconcile() error string = '%v', errorString = '%v'", err, tt.errorString)
+			}
```
I did something a bit more consistent. FWIW I extremely dislike the existing practice because it makes knowing what went on when it's not your machine (e.g. CI) harder.

m1kola reviewed (changes requested):
Could you please open work items/separate PRs for these 3 topics?

* https://github.com/Azure/ARO-RP/pull/1845#discussion_r806681619
* https://github.com/Azure/ARO-RP/pull/1845#discussion_r806710313
* https://github.com/Azure/ARO-RP/pull/1845#discussion_r809437881

IMO they are not blocking and can be addressed separately, but I want to make sure that we track them before we merge this PR.

I left few comments in some other existing discussions. We still need to figure out:

* What and how we reconcile [here](https://github.com/Azure/ARO-RP/pull/1845#discussion_r812065813) and [here](https://github.com/Azure/ARO-RP/pull/1845#discussion_r812073913)
* And change the controller test to make it more maintanable. I left [this note](https://github.com/Azure/ARO-RP/pull/1845#discussion_r812189539) on how I see it in my previous review, but because of my fat fingers you missed it (I resolved the convo accidentally)

Everything else looks good to me.

m1kola reviewed (changes requested):
I think this is almost ready to merge. I like how controller and deployer ended up looking. Controller tests are also much more readable in my opinion: being familiar with go-mock it is very easy to understand what test case covers what. Deployer tests are not ideal, but I think we found a good balance between testing too much and testing too little.

I have few comments about latest changes to dynamic helper: most of them are just questions, but it looks like we are missing test coverage for new changes.

---


> Could you please open work items/separate PRs for these 3 topics?
> 
> * https://github.com/Azure/ARO-RP/pull/1845#discussion_r806681619
> * https://github.com/Azure/ARO-RP/pull/1845#discussion_r806710313
> * https://github.com/Azure/ARO-RP/pull/1845#discussion_r809437881
> 
> IMO they are not blocking and can be addressed separately, but I want to make sure that we track them before we merge this PR.

I see that we solved [`GENERATED` issue](https://github.com/Azure/ARO-RP/pull/1845#discussion_r806710313) locally, but we still might want to contribute to the upstream MUO so we don't end up with a quay image again after copying yamls from the MUO's next release.

hawkowl commented on pkg/util/dynamichelper/dynamichelper.go:
```diff
@@ -210,6 +223,17 @@ func merge(old, new kruntime.Object) (kruntime.Object, bool, string, error) {
 		old, new := old.(*arov1alpha1.Cluster), new.(*arov1alpha1.Cluster)
 		new.Status = old.Status
+
+	case *corev1.ConfigMap:
+		old, new := old.(*corev1.ConfigMap), new.(*corev1.ConfigMap)
+
+		_, injectTrustBundle := new.ObjectMeta.Labels["config.openshift.io/inject-trusted-cabundle"]
+		if injectTrustBundle {
+			caBundle, ext := old.Data["ca-bundle.crt"]
+			if ext {
+				new.Data["ca-bundle.crt"] = caBundle
+			}
+		}
```
Yes, dynamichelper will copy some elements from the existing copy (e.g. the scc things) and then PUT it.

hawkowl commented on hack/validate-imports/validate-imports.go:
```diff
@@ -209,6 +214,7 @@ func validateImports(path string, fset *token.FileSet, f *ast.File) (errs []erro
 		"pkg/database/cosmosdb/zz_generated_",
 		"pkg/operator/apis",
 		"pkg/operator/clientset",
+		"pkg/operator/mocks/",
```
I wasn't so sure about putting Operator controller specific mocks into the util directory, just in case the Operator ends up being pulled out/separated/etc as a part of future things. Similar to how the API is kept in the operator, not with the rest of them.

hawkowl commented on pkg/util/dynamichelper/dynamichelper.go:
```diff
@@ -173,6 +174,12 @@ func merge(old, new kruntime.Object) (kruntime.Object, bool, string, error) {
 			copyAnnotation(&new.ObjectMeta, &old.ObjectMeta, name)
 		}
+		// Copy OLM label
+		for k, v := range old.Labels {
```
done

hawkowl commented:
I ran out of time to add a test for the appsv1.Deployment changes in DynamicHelper since it adds a bunch of default stuff and I'm not 100% sure how I can make the existing test happy with that without copy-pasting a bunch of defaults that aren't actually relevant to the test. Added tests for the other items in it, though, and fixed a potential crash (yay for tests!).

m1kola reviewed (approved):
Looks great, thanks @hawkowl! I'm going to merge this PR, but please remember to create work items for remaining things mentioned in the [previous reviews](https://github.com/Azure/ARO-RP/pull/1845#pullrequestreview-892309699).


---

> I'm not 100% sure how I can make the existing test happy with that without copy-pasting a bunch of defaults that aren't actually relevant to the test. 

`scheme.Scheme.Default(old)` and/or `scheme.Scheme.Default(want)` should help here.

I created [this work item](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/13525478) to increase test coverage in `merge` and in `dynamichelper` in general. I see that a lot of existing cases are still need coverage.
